### PR TITLE
 Add reference counting to buffer pool

### PR DESCRIPTION
**Pull request by ganwell**
The user may release a message after the connection has shutdown.
So we must keep the buffer pool around till the last message is released.

Attention: Rebase needed - only review last commit

### CODE REVIEW

**Comment by ganwell** [comment]
@winged rebased, ready to merge

File: src/buffer.c [hunk 1]
```diff
@@ -73,6 +77,7 @@ ch_bf_init(ch_buffer_pool_t* pool, ch_connection_t* conn, uint8_t max_buffers)
     A(max_buffers <= 32, "buffer.c can't handle more than 32 handlers");
     memset(pool, 0, sizeof(*pool));
     pool->conn         = conn;
+    pool->refcnt       = 1;
```
Comment (winged) [inline src/buffer.c]: I'm not sure why you initialize `refcnt` with 1.. shouldn't this be 0 at first?

File: src/buffer.c [hunk 2]
```diff
@@ -110,6 +115,7 @@ ch_bf_acquire(ch_buffer_pool_t* pool)
     if (pool->used_buffers < pool->max_buffers) {
         int free;
         pool->used_buffers += 1;
+        pool->refcnt += 1;
```
Comment (winged) [inline src/buffer.c]: Just to confirm: Are we sure this only happens in one thread, ever? Otherwise we probably need to lock this

Comment (ganwell) [reply src/buffer.c]: Good point, the pool->refcnt -= 1 can happen in an other thread if user ignores our advice, ch_chirp_release_message is not marked as thread-safe. We cant make ch_chirp_release_message thread-safe by just adding locks, would have to use uv_async_t to make a ch_chirp_release_message_ts variant, which we might do in the future. 

TLDR: it is corrent. But good point.

File: src/buffer.c (hunk 1 above)
Comment (ganwell) [reply src/buffer.c]: The pool is initially owned by the reader (of the connection). So

1. No need to make things complicated and make the owner count up, there will always be an owner, or it is bug.
2. The name refcnt is may be not perfect, its more own_plus_refcnt
